river/View.zig

```diff
@@ -379,3 +402,73 @@ pub fn unmap(self: *Self) void {
 
     root.startTransaction();
 }
+
+/// Change the opacity of a view by config.view_opacity_delta.
+/// If the target opacity was reached, return true.
+fn incrementOpacity(self: *Self) bool {
+    // TODO damage view when implementing damage based rendering
+    const config = self.output.root.server.config;
+    if (self.opacity < self.current.target_opacity) {
+        self.opacity += config.view_opacity_delta;
+        if (self.opacity < self.current.target_opacity) return false;
+    } else {
+        self.opacity -= config.view_opacity_delta;
+        if (self.opacity > self.current.target_opacity) return false;
+    }
+    self.opacity = self.current.target_opacity;
+    return true;
+}
+
+/// Destroy a views opacity timer
+fn killOpacityTimer(self: *Self) void {
+    if (c.wl_event_source_remove(self.opacity_timer) < 0) unreachable;
+    self.opacity_timer = null;
+}
+
+/// Set the timeout on a views opacity timer
+fn armOpacityTimer(self: *Self) void {
+    const delta_t = @intCast(c_int, self.output.root.server.config.view_opacity_delta_t);
+    if (c.wl_event_source_timer_update(self.opacity_timer, delta_t) < 0) {
+        log.err(.view, "failed to update opacity timer", .{});
+        self.killOpacityTimer();
+    }
+}
+
+/// Called by the opacity timer
+fn handleOpacityTimer(data: ?*c_void) callconv(.C) c_int {
+    const self = util.voidCast(Self, data.?);
+    if (self.incrementOpacity()) {
+        self.killOpacityTimer();
+    } else {
+        self.armOpacityTimer();
+    }
+    return 0;
+}
+
+/// Create an opacity timer for a view and arm it
+fn attachOpacityTimer(self: *Self) void {
+    const server = self.output.root.server;
+    self.opacity_timer = c.wl_event_loop_add_timer(
+        c.wl_display_get_event_loop(server.wl_display),
+        handleOpacityTimer,
+        self,
+    ) orelse return;
+    self.armOpacityTimer();
+}
+
+/// Start an opacity transition
+pub fn startOpacityTransition(self: *Self, target_opacity: f32) void {
+    self.pending.target_opacity = target_opacity;
+}
```

Comment:
I'm not really a fan of one line functions like this. IMO they don't really pull their weight in terms of readability as I always end up going to the definition to see what it actually does.